### ISSUE TITLE
npm 6 does not update git dependencies with SHA in package-lock.json

### ISSUE THREAD

**Issue by targos** [npm/npm#20485]
#### I'm opening this issue because:

  - [x] npm is producing an incorrect install.

#### What's going wrong?

I have a git dependency in a package, pinned to a commit SHA. If I change the SHA in my `package.json`, the `package-lock.json` stays on the previous SHA and the dependency is not updated.
This worked as expected in npm 5.

#### How can the CLI team reproduce the problem?

- Clone `https://github.com/targos/bug-npm6.git`
- Run `npm install`

This will install `github:targos/test-package#3a31a39f1f8d45798cca5079531d8b5c344f8669` (the version in `package-lock.json`) instead of `github:targos/test-package#3c0be7d6f276cd1ccb56110bce941abc9ffcd837` (the version in `package.json`)

### supporting information:

 - `npm -v` prints: 6.0.0
 - `node -v` prints: 10.0.0
 - `npm config get registry` prints: https://registry.npmjs.org/
 - Windows, OS X/macOS, or Linux?: Windows and Linux
 - Network issues:
   - Geographic location where npm was run: Switzerland

**Comment by mivion** [comment]
Same problem...

**Comment by motet-a** [comment]
I have a similar problem with npm 6 and the `git-https` protocol with `x-oauth-basic` (with a secret token in the URL) and a git tag. When I change the git tag in the `package.json`, the `package-lock.json` and the installed package are not updated.

The dependency entry in my `package.json` looks like this:

```json
"some-package": "git+https://secretaccesstoken67896789123:x-oauth-basic@github.com/foo-org/bar-repo#some-git-tag",
```

Everything works fine with npm 5.

**Comment by motet-a** [comment]
In fact, in both cases, npm 5.10.0 is affected too. 5.8.0 is not affected.

**Comment by motet-a** [comment]
I just found with `git bisect` the commit that introduces the bug: https://github.com/npm/npm/commit/5f84ebdb66e35486d1dec1ca29e9ba0e4c5b6d5f

Hope this helps.

**Comment by iarna** [comment]
This was an intentional change at the time (and I believe fresh lockfiles generated at the time of 5f84ebd didn't demonstrate this bug, though it does with current lockfiles) but since adding `from` we can tighten up and simplify the code:

```diff
8538ea038 deps: Now that from is in the lockfile, tighten git matching (Rebecca Turner on 2018-05-18)
diff --git a/lib/install/deps.js b/lib/install/deps.js
index d03c1b9f2..f89a4a9d1 100644
--- a/lib/install/deps.js
+++ b/lib/install/deps.js
@@ -54,11 +54,7 @@ function doesChildVersionMatch (child, requested, requestor) {
   if (requested.type === 'git' && child.fromShrinkwrap) {
     const fromSw = child.fromShrinkwrap
 
-    if (fromSw.rawSpec === requested.rawSpec) return true
-
-    if (!fromSw.hosted || !requested.hosted) return false
-
-    return fromSw.hosted.toString() === requested.hosted.toString()
+    return fromSw.toString() === requested.toString()
   }
 
   if (!registryTypes[requested.type]) {
```

(This'll be going out soon.)

**Comment by targos** [comment]
This is fixed in npm@6.1.0. Thank you!
